cmd/cstorvolume-claim/cvc/cstorvolumeclaim.go

```diff
@@ -0,0 +1,399 @@
+		Get(claim.Name, metav1.GetOptions{})
+
+	if err != nil && !k8serror.IsNotFound(err) {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get cstorvolume service {%v}",
+			svcObj.Name,
+		)
+	}
+	if k8serror.IsNotFound(err) {
+		svcObj, err = svc.NewBuilder().
+			WithName(claim.Name).
+			WithLabelsNew(labels).
+			WithAnnotations(annotations).
+			WithOwnerRefernceNew(OwnerReference).
+			WithSelectorsNew(selectors).
+			WithPorts(ports).
+			Build()
+		if err != nil {
+			return nil, errors.Wrapf(
+				err,
+				"failed to build target service {%v}",
+				svcObj,
+			)
+		}
+
+		svcObj, err = svc.NewKubeClient(svc.WithNamespace("openebs")).Create(svcObj)
+	}
+	return svcObj, err
+}
+
+// cstorvolumereplicacr creates CStorVolume resource for a cstor volume
+func createCStorVolumecr(service *corev1.Service,
+	claim *apis.CStorVolumeClaim,
+	scName string,
+) (*apis.CStorVolume, error) {
+
+	labels := map[string]string{
+		"openebs.io/target-service":    "cstor-target-svc",
+		"openebs.io/persistent-volume": claim.Name,
+	}
+	annotations := map[string]string{
+		//"openebs.io/storage-class-ref": "name: " + storageClassName,
+		"openebs.io/fs-type": "ext4",
+		"openebs.io/lun":     "0",
+	}
+	OwnerReference := []metav1.OwnerReference{
+		*metav1.NewControllerRef(claim,
+			apis.SchemeGroupVersion.WithKind("CStorVolumeClaim")),
+	}
+
+	qCap := claim.Spec.Capacity[corev1.ResourceStorage]
+	rfactor, cfactor, err := getReplicationFactor(scName)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get replica-count and factor {%s}",
+			scName,
+		)
+	}
+
+	cvObj, err := cv.NewKubeclient(cv.WithNamespace("openebs")).
+		Get(claim.Name, metav1.GetOptions{})
+	if err != nil && !k8serror.IsNotFound(err) {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get cstorvolume {%v}",
+			cvObj.Name,
+		)
+	}
+	if k8serror.IsNotFound(err) {
+		cvObj, err = cv.NewBuilder().
+			WithName(claim.Name).
+			WithLabelsNew(labels).
+			WithAnnotationsNew(annotations).
+			WithOwnerRefernceNew(OwnerReference).
+			WithTargetIP(service.Spec.ClusterIP).
+			WithCapacity(qCap.String()).
+			WithCStorIQN(claim.Name).
+			WithTargetPortal(service.Spec.ClusterIP + ":" + "3260").
+			WithTargetPort("3260").
+			WithReplicationFactor(rfactor).
+			WithConsistencyFactor(cfactor).
+			Build()
+		if err != nil {
+			return nil, errors.Wrapf(
+				err,
+				"failed to get cstorvolume {%v}",
+				cvObj,
+			)
+		}
+
+		return cv.NewKubeclient(cv.WithNamespace("openebs")).Create(cvObj)
+	}
+	return cvObj, err
+}
+
+// createCStorVolumeReplica create cstorvolume replica based on the replicaCount
+// on the available cstor pools matched with storagepool claim given in
+// storageClass as parameter value.
+// if pools are less then replicaCount we return with error.
+func createCStorVolumeReplica(
+	service *corev1.Service,
+	volume *apis.CStorVolume,
+	scName string,
+) error {
+
+	replicaCount, spcName, err := getFromParameters(scName)
```

Comment:
this function name doesn't sounds good.. probably, pass the scObj instead of scName, and get the required fields from scObj directly